File: src/main/java/com/github/iamniklas/liocore/procedures/variants/LightningProcedure.java
package com.github.iamniklas.liocore.procedures.variants;

import com.github.iamniklas.colorspaces.ColorRGB;
import com.github.iamniklas.liocore.led.LEDDataBundle;
import com.github.iamniklas.liocore.led.colorspace.LIOColor;
import com.github.iamniklas.liocore.network.LEDUpdateModel;
import com.github.iamniklas.liocore.procedures.Procedure;

public class LightningProcedure extends Procedure {

    public LightningProcedure(LEDUpdateModel _ledUpdateModel) {
        super(_ledUpdateModel);
    }

    @Override
    public void start() {
        super.start();

        steps = ledUpdateModel.bundle.duration * 50;
    }

    @Override
    public void update() {
        strip.setAllPixels(LIOColor.fromRGB(ColorRGB.WHITE50.dim((float) Math.abs(Math.sin((step / ledUpdateModel.bundle.speed))))));

        ledUpdateModel.bundle.speed += ledUpdateModel.bundle.attenuation;

        if(step > steps) {
            finishProcedure(true);
        }
        step++;
    }

    @Override
    public void updateLEDDataBundle(LEDDataBundle ledDataBundle) {

    }

    @Override
    public boolean validateBundleData() {
        return ledUpdateModel.bundle.speed != null && ledUpdateModel.bundle.attenuation != null && ledUpdateModel.bundle.duration != null;
    }
}

File: src/main/java/com/github/iamniklas/liocorekotlin/procedures/LightningProcedure.kt
package com.github.iamniklas.liocorekotlin.procedures

import com.github.iamniklas.liocorekotlin.led.ColorRGB
import com.github.iamniklas.liocorekotlin.led.LEDDataBundle

class LightningProcedure(_bundle: com.github.iamniklas.liocorekotlin.led.LEDDataBundle) : com.github.iamniklas.liocorekotlin.procedures.Procedure(_bundle) {
    private var mSpeed: Float
    private val mAtenuation: Float
    override fun start() {}
    public override fun update() {
        mStrip!!.setAllPixels(
            com.github.iamniklas.liocorekotlin.led.ColorRGB.white50.dim(Math.abs(Math.sin((mStep / mSpeed).toDouble())).toFloat()).toSystemColor()
        )
        mSpeed += mAtenuation
        if (mStep > mSteps) {
            finishProcedure(true)
        }
        mStep++
    }

    init {
        mSpeed = _bundle.speed!!
        mAtenuation = _bundle.value1!!
        mSteps = _bundle.duration!! * 50
    }
}